### ISSUE TITLE
Windows IA-64 guest support

### ISSUE THREAD

**Issue by binarymaster** [trofi/ski#5]
I wonder if this software could support Windows for IA-64 as guest system one day, at least [Windows XP 64-Bit Edition](https://betawiki.net/wiki/Windows_XP_64-Bit_Edition) that targeted original Itanium / Merced processor.

Recently I also found out that HP-UX could run such guests virtualized: https://virtuallyfun.com/2024/09/23/windows-server-2008-ia64-guest-on-hp-integrity-vm-hp-ux/ - maybe this information could help achieve emulation of Windows for Itanium-based systems.

**Comment by trofi** [comment]
It's probably not easy to run `windows` as a guest. Today `ski` relies on a few things to get something running: modified guest kernel has to to pass data in and out for the console output (for example), to emulate block device (instead of simulating real bus like PCI), there is no real ACPI namespace exported to the guest. No video card support either.

**Comment by BHTY** [comment]
It could probably be more feasible to use Ski to run Win64/IA-64 applications while not actually booting IA-64 Windows, instead using the CPU emulator while thunking up API calls, in a similar manner to my EmuWOW project.

**Comment by sskras** [comment]
Some related project (of userland emulation) could be found here: @itanium64 | https://itanium64.github.io/

It lived through two rewrites: Go –> Rust –> C#, only to get abandoned, it seems.

**Comment by Vincent392** [comment]
> It could probably be more feasible to use Ski to run Win64/IA-64 applications while not actually booting IA-64 Windows, instead using the CPU emulator while thunking up API calls, in a similar manner to my EmuWOW project.

By the time you would even be able to GET there, ski would probably be *able* to load the IA-64 NTOSKRNL anyways and boot the whole OS as making Ski into an app layer would require a whole rewrite.

**Comment by NexusSfan** [comment]
You could try [Simics](https://en.wikipedia.org/wiki/Simics) if you don't mind Proprietary software

**Comment by Vincent392** [comment]
Returning to this, I tried for the hell of it to boot IA-64 Windows PE, and *this* returned

```
bski -forcesystem -stats -noconsole /media/random/WinGlass/InstallerIA64/WinGlassPE/IA64/setupldr.efi
Halting Simulation
119 insts (1 fault), 0.00 sec, 1166667 i/s, 31 cycles, 3.84 ipc
```

It DOES start very briefly from the seems of it, however it obviously wont work properly since it's using just the raw setupldr file, not an ISO.

It's not out of the question.

**Comment by Vincent392** [comment]
@BHTY I assume it's because of the fact it's running right from the source folder and NOT the ISO the PE expects itself to be booting from.

That, or it's searching for TXTSETUP.SIF and the drivers and panicking because it can't.

**Comment by sskras** [comment]
@Vort commented [34 minutes ago](https://github.com/trofi/ski/issues/5#issuecomment-4183505113):

> Thanks for explanation. I can't understand, however, why it may be better to implement ATI ES1000 first. Why not get VGA output first (with random card ported from somewhere), make Windows work reliably (with basic video capabilities) and then proceed with implementation of more advanced, 3D accelerated video options?

I guess that was suggested because it's a pretty small set of GPUs that were supported on Windows-ia64 architecture + we don't have [an expensive gear](https://github.com/trofi/ski/issues/5#issuecomment-4136049494) to test GPU cards by hand.

@Vort commented [10 hours ago](https://github.com/trofi/ski/issues/5#issuecomment-4181659987):

> > The cheapest Merced machine I could find is HP i2000 Workstation running Itanium 733MHz + 16GB RAM + NVIDIA Quadro2 Pro
> 
> Does anyone knows what will happen if user installs different GPU there - GeForce 3 or Voodoo 3 for example? At which point it will fail?

GPUs up to Geforce 5 might be supported on that machine according to this Reddit thread, mentioning HP driver packages: [**512mb GPUs and Itanium?**](https://www.reddit.com/r/homelab/comments/te8jz6/512mb_gpus_and_itanium/#:~:text=looks%20like,(including%20Geforce%205%20series)&text=are%20all%20supported)

> > Both ATI and Nvidia GPUs were supported by Windows for Itanium. From what I can tell, the drivers were commissioned by HP themselves. I have the HP softpaqs for numerous driver versions and from what it looks like, Nvidia GPUs up to the Quadro FX 3000 (including Geforce 5 series) and ATI FireGL X1, Z1, and X2 cards are all supported.

So this gives some hope.

> Will it display at least something during boot? Will it be possible to boot Windows in safe mode or with generic (VESA, VGA) driver?

According to the VCFed forum post I mentioned above:

> > Microsoft and their infinite wisdom didn't keep VESA modes, which were far faster.

So I doubt about VESA support here.

>  If Voodoo 3 will do at least something, it may help in developing GPU emulation, since there are many different implementations for it available for porting.

Having no experience with running anything graphical on Itanium, I didn't even know there was [VooDoo 3 drivers for Itanium](https://download.cnet.com/3dfx-voodoo3/3000-2108_4-152748.html#:~:text=Windows%20XP-,Itanium%2064%2Dbit,-Windows%20NT%204&text=Windows%202003-,Itanium%2064%2Dbit,-Windows%20NT%204) builds! That's pretty fascinating.

**Comment by sskras** [comment]
PS. What a nice piece available for £1,741.04 from Newtongrange, UK on eBay:
[**hp workstation zx6000 - Dual Itanium 2 1.5 GHz - 8 GB RAM**](https://www.ebay.co.uk/itm/235081713025)

**Comment by sskras** [comment]
I am amazed by how rare is the mention of IA-64 build of nVidia ForceWare drivers on the todays web.

There is a [**ForceWare Release 50 Graphics Drivers User’s Guide** | **Version 52.16 for Windows**](https://ftp.braindrainlan.nu/Drivers/Graphic%20Cards/NVIDIA/52.16/ForceWare_Windows_XP_Release_50_graphics_drivers_Display_Properties_Users_Guide_v52.16.pdf) which says on page 13:

<img  width="75%" alt="Image" src="https://github.com/user-attachments/assets/ba51c721-da60-4cf2-9758-9f4fa347bcab"/> <br />

But the only versions explicitly mentioning `IA-64` on their download pages that I managed to dig up using Google are:

* [nVidia Forceware 57.30 64-bit](https://tweakers.net/downloads/5773/nvidia-forceware-5730-64-bit.html) => [`57.30_winxp64_international.exe`](https://web.archive.org/web/*/http://download.nvidia.com/Windows/64/57.30/57.30_winxp64_international.exe)
* [ForceWare 78.03 Windows x64](https://bestfilez.net/forums/index.php?showtopic=635&st=200#entry205115) => [`78.03_winxp64_international.exe`](https://web.archive.org/web/*/ftp://download1.nvidia.com/Windows/64/78.03/78.03_winxp64_international.exe)
* [ForceWare 77.50 Windows XP 64-bit](https://www.guru3d.com/download/forceware-77-50-windows-xp-64-bit/) => [`7750-x64-whql-[Guru3D.com].exe`](https://ftp.nluug.nl/pub/games/PC/guru3d/nvidia/7750-x64-whql-[Guru3D.com].exe)

And even with this I **fail** to find the `ia64` section in their `nv4_disp.inf` files. Does the section need to exist? 

It seems not, after reading this post [**INF File Installation Section for 64-Bit Architecture**](https://community.osr.com/t/inf-file-installation-section-for-64-bit-architecture/31059/3#:~:text=Of%20course%20you%20can%20use%20one%20INF%20file%20for%20both%20platforms%2C%20using%20decorated%0AINF%20sections) from the OSR Developer Community. But I am still unsure if the packages above contain any IA-64 binaries...

---

The only place that led me to the seemingly real IA-64 drivers for Windows was the [**Need help modifying IA-64 Nvidia drivers**](https://forums.guru3d.com/threads/need-help-modifying-ia-64-nvidia-drivers.442138/) post on Guru3D forums. Surprisingly it was the same package SP28071 package provided by HP that @Vort mentioned in the last comment. Later I discovered older versions too. Listing them in chronological order:

* [zx1 AGP Services driver v.5.2.1.0 Rev. A](https://driverscollection.com/?file_cid=44758376520834531840998d6ee) => `sp24283.exe`
* [NVIDIA Unified Driver v.50.21 Rev. A](https://driverscollection.com/?file_cid=447583751349d69512d681cffaa) => `sp25515.exe` (56 Device Ids)
* [NVIDIA Unified Driver v.53.02 Rev. A](https://driverscollection.com/?file_cid=4475837539834db3c14f4fa5bd5) => `sp25945.exe` (4 Device Ids, 2 new)
* [NVIDIA IA-64 Graphics Driver v.53.18 Rev. 1](https://driverscollection.com/?file_cid=44758377510389c1573dc2899a7) => `sp28071.exe` (3 Device Ids, all of them repeat)

<details> <summary> The extracted <code>SP28071.txt</code> </summary> <br />

```txt
TITLE: NVIDIA IA-64 Graphics Driver for Itanium Workstations

VERSION: 53.18 Rev. 1
DESCRIPTION:
This contains the NVIDIA graphics driver for the Itanium workstation models and 
operating systems listed. 

PURPOSE: Routine
SOFTPAQ NUMBER: SP28071
SUPERSEDES: None
EFFECTIVE DATE: June 10, 2004
CATEGORY: Driver - Graphics - Nvidia
SSM SUPPORTED: No

PRODUCT TYPE(S):
Workstations

PRODUCT MODEL(S):
HP Workstation zx2000: zx2000,zx6000
HP Workstation zx2000: zx2000,zx6000
HP Workstation zx6000: zx2000,zx6000
HP Workstation zx6000: zx2000,zx6000

OPERATING SYSTEM(S): 
Windows XP 64 Itanium

LANGUAGE(S): 
US English

ENHANCEMENTS: 
Latest driver release for IPF.  Supports latest driver features from nVIDIA.

PREREQUISITES: 
Must have the HP zx1 AGP GART driver installed (version 5.2.1.0 Rev A, Softpaq 
number: SP24283).

HOW TO USE: 
1. Download the SoftPaq to a directory on your hard drive.  The file that is 
downloaded is a self-extracting  executable with a filename based on the 
SoftPaq Number above. 

2. From that drive and directory, execute the downloaded file and follow the 
on-screen instructions.



Copyright (c) 2004 Hewlett-Packard Development Company, L.P.
```
</details>

Then since it said it installed on my w10 21H2 (64-bit) successfully, I am puzzled about how the IA-64 driver setup works internally.

The page has other _zx2000_ goodies, eg:
* [ATI Radeon 7000 Video Graphics Driver v.6.13.10.6118](https://driverscollection.com/?file_cid=44758376223ca108d4bde5f0513) => `ativideo6118.zip` (1 Device Id)
* [ATI FireGL X1/­Z1 Driver v.3763](https://driverscollection.com/?file_cid=44758375761dc92324b6407873d) => `sp24508.exe` (4 Device Ids)
* [ATI FireGL Driver v.7.91.90919 Rev. A](https://driverscollection.com/?file_cid=44758374903a04d57ba57e38b25) => `sp25944.exe` (8 Device Ids)
* [ATI FireGL Driver (IA-64 Itanium) v.8.02.3 Re. A](https://driverscollection.com/?file_cid=44758378500eb13134b0b4e9395) => `sp28437.exe` (8 Device Ids)

In case someone hits the download limit, here is two mirrors: 
https://drivers.eu/PC/HP/zx2000
https://ftp.zx.net.nz/pub/archive/ftp.compaq.com/pub/softpaq/

Now circling back to nVidia on Itanium setup, the post points to another issue:

* [HP zx2000 Workstation BIOS v.5.00 Rev 1](https://fr.driverscollection.com/_47293447311dbb06b973bb3b594/T%C3%A9l%C3%A9charger-HP-zx2000-Workstation-ATI-FireGL-Pilote-(IA-64-Itanium)-v.8.02.3-Re.-A-pour-Windows-XP-64-bit-gratuit) => `sp27252.exe`

> ### BIOS for HP zx2000 Workstation
> This contains the EFI version of the NVIDIA Video BIOS Flash
> Utility for the Itanium-based (IA-64) workstations listed. This
> flash utility may be used with the same NVIDIA video BIOS images
> that are provided for x86 systems

<details> <summary> The extracted <code>SP27252.txt</code> </summary> <br />

```txt
TITLE: NVIDIA EFI Video BIOS Flash Utility for Itanium-based Workstations

VERSION: 5.00 Rev. 1
DESCRIPTION:
This contains the EFI version of the NVIDIA Video BIOS Flash Utility for the 
Itanium-based (IA-64) workstations listed. This flash utility may be used with 
the same NVIDIA video BIOS images that are provided for x86 systems. 

PURPOSE: Recommended                                                     
SOFTPAQ NUMBER: SP27252
SUPERSEDES: None
EFFECTIVE DATE: February 16, 2004
CATEGORY: EFI (extensible firmware interface)                             
SSM SUPPORTED: No

PRODUCT TYPE(S):
Workstations

PRODUCT MODEL(S):
HP Workstation zx2000: All Models
HP Workstation zx2000: All Models
HP Workstation zx6000: All Models
HP Workstation zx6000: All Models

OPERATING SYSTEM(S): 
Microsoft Windows 2003 Server
RH Ent Linux 3 (Itanium)
RH Ent Linux WS 2.1 (Itanium)
Windows XP 64

LANGUAGE(S): 
US English

ENHANCEMENTS: 
N/A

PREREQUISITES: 
N/A

HOW TO USE: 
1. Download the SoftPaq to a directory on your hard drive. The file downloaded 
is a zip compressed file that will extract to a 25MB ISO image that can be 
burned on a CD and booted in EFI.

2. Burn the extracted ISO image onto a CD using a CD burner.

3. Insert the EFI bootable CD you just burned in the previous step into the ROM 
drive of the IPF system with the NVIDIA product.

4. Boot to the EFI shell and change directory to the CDROM filesystem (Usually 
the last 'fs#:' - for example, 'fs2:' - on zx2000 or 'fs0:' on the zx6000).

5. If you type the command 'dir' you should see a script file called 
'nvflash.efi'. Otherwise, try switching to another drive until you locate it.

6. Once you locate the filesystem with the 'nvflash.efi' file, you may copy it 
to the EFI partition along with any NVIDIA video BIOS image you may have.

7. To run it, simply type 'nvflash.efi ' {where is the name of a valid NVIDIA 
video BIOS image file.}

7. Once the video BIOS update is complete, reboot the system {you can do a soft 
reboot by typing the command 'reset'}.



Copyright (c) 2004 Hewlett-Packard Development Company, L.P.
```
</details>

It's still unclear for me if the IA-64 specific VBIOS firmware is needed for the GeForce cards to boot on Itanium architecture.

There are folks saying it's needed: [**Trying to find official Itanium Nvidia VBIOS**](https://www.reddit.com/r/homelab/comments/wq2o3v/comment/iklbp62/)

> [**xiyatumerica**](https://www.reddit.com/user/xiyatumerica/) OP • [4y ago](https://www.reddit.com/r/homelab/comments/wq2o3v/comment/iklbp62/)
> 
> > The itanium processor isn't initializing the card because it has an x86 BIOS

> [**frankd412**](https://www.reddit.com/user/frankd412/) • [4y ago](https://www.reddit.com/r/homelab/comments/wq2o3v/comment/ikny4x4/)
> 
> > You mean the option rom? Does it matter? Do you need video before the OS?

> > [**xiyatumerica**](https://www.reddit.com/user/xiyatumerica/) OP • [4y ago](https://www.reddit.com/r/homelab/comments/wq2o3v/comment/iko3vxm/)
> > 
> > > It needs to be the vBIOS because the itanium 2 can't itialize the device (since it's running an x86 bios)

There are also folks saying UGA/GOP firmware is not required on most IA-64 platforms:
[**Which Nvidia GPUs are compatible with Itanium cpus?**](https://www.reddit.com/r/nvidia/comments/m1mdjr/which_nvidia_gpus_are_compatible_with_itanium_cpus/)

> [**yuhong**](https://www.reddit.com/user/yuhong/) • [5y ago](https://www.reddit.com/r/nvidia/comments/m1mdjr/comment/gqnndhm/)
> > I believe that most IA64 platforms did not require UGA/GOP firmware on the video cards, and in fact Windows required VGA/VESA support.
> 

> > [**yuhong**](https://www.reddit.com/user/yuhong/) • [5y ago](https://www.reddit.com/r/nvidia/comments/m1mdjr/comment/gqnpetq/)
> > 
> > > Trivia: the Windows /NOVESA switch was targeted at HP IA64 servers: http://h10032.www1.hp.com/ctg/Manual/c00090141.pdf
> 

Discussions like this leaves me even more puzzled.

PS. Found some screenshots from flashing the UEFI/GOP image, but this seems to be unrelated to IA-64 – more like dedicated to the VBIOS modding on modern nVidia cards: [**LEGACY/UEFI Boot BIOS settings**](https://winraid.level1techs.com/t/legacy-uefi-boot-bios-settings/30762/111)

**Comment by Vort** [comment]
> It then returned with `loading setupldr.efi...` before seemingly freezing on the frontend.

> Hell, maybe it's even fully booted into the PE, but we don't know

Source code of `ski-efi` shows that after such message, `starting kernel...` message should appear. Its absence means hang in EFI loader.

In case of `setupldr.efi`, hang happens on this line:
https://github.com/trofi/ski/blob/6984f89d0d529c8b752f03d3d623f57033d00e58/ski-efi/bootloader.c#L231-L232
because `sec_hdr->raw_data_size` is larger than `sec_hdr->virtual_size`.
Adding `if (sec_hdr->virtual_size > sec_hdr->raw_data_size)` before this line allows loading to go further. Now it hits
https://github.com/trofi/ski/blob/6984f89d0d529c8b752f03d3d623f57033d00e58/ski-efi/bootloader.c#L286
Does anyone know how to implement missing cases there?

**upd:** I made separate issue for this problem: #29.

**Comment by Vincent392** [comment]
> > It then returned with `loading setupldr.efi...` before seemingly freezing on the frontend.
> 
> > Hell, maybe it's even fully booted into the PE, but we don't know
> 
> Source code of `ski-efi` shows that after such message, `starting kernel...` message should appear. Its absence means hang in EFI loader.
> 
> In case of `setupldr.efi`, hang happens on this line:
> https://github.com/trofi/ski/blob/6984f89d0d529c8b752f03d3d623f57033d00e58/ski-efi/bootloader.c#L231-L232
> because `sec_hdr->raw_data_size` is larger than `sec_hdr->virtual_size`.
> Adding `if (sec_hdr->virtual_size > sec_hdr->raw_data_size)` before this line allows loading to go further. Now it hits
> https://github.com/trofi/ski/blob/6984f89d0d529c8b752f03d3d623f57033d00e58/ski-efi/bootloader.c#L286
> Does anyone know how to implement missing cases there?
> 
> **upd:** I made separate issue for this problem: #29.

We're getting somewhere holy shit

**Comment by Vincent392** [comment]
Here's a practically unmodified XP IA-64 PR

https://archive.org/download/cursed_xpwinpes/IA-64/en_windows_glass_professional_ia64_dvd_vl_m3.iso

Have fun trying to boot it
